src/data/polynomial/laurent.lean

```diff
@@ -0,0 +1,168 @@
+add_monoid_algebra.single_zero_ring_hom
+
+lemma single_eq_C (r : R) : single 0 r = C r := rfl
+
+/--  The function `n ↦ T ^ n`, implemented as a sequence `ℤ → R[T;T⁻¹]`.
+
+Using directly `T ^ n` does not work, since we want the exponents to be of Type `ℤ` and there
+is no `ℤ`-power defined on `R[T;T⁻¹]`.  Using that `T` is a unit introduces extra coercions.
+For these reasons, the definition of `T` is as a sequence. -/
+def T (n : ℤ) : R[T;T⁻¹] := single n 1
+
+@[simp]
+lemma T_zero : (T 0 : R[T;T⁻¹]) = 1 := rfl
+
+lemma T_add (m n : ℤ) : (T (m + n) : R[T;T⁻¹]) = T m * T n :=
+by { convert single_mul_single.symm, simp [T] }
+
+@[simp]
+lemma T_pow (m : ℤ) (n : ℕ) : (T m ^ n : R[T;T⁻¹]) = T (n * m) :=
+by rw [T, T, single_pow n, one_pow, nsmul_eq_mul, int.nat_cast_eq_coe_nat]
+
+/-- The `simp` version of `mul_assoc`, in the presence of `T`'s. -/
+@[simp]
+lemma mul_T_assoc (f : R[T;T⁻¹]) (m n : ℤ) : f * T m * T n = f * T (m + n) :=
+by simp [← T_add, mul_assoc]
+
+@[simp]
+lemma single_eq_C_mul_T (r : R) (n : ℤ) :
+  (single n r : R[T;T⁻¹]) = (C r * T n : R[T;T⁻¹]) :=
+by convert single_mul_single.symm; simp
+
+-- This lemma locks in the right changes and is what Lean proved directly.
+-- The actual `simp`-normal form of a Laurent monomial is `C a * T n`, whenever it can be reached.
+@[simp]
+lemma _root_.polynomial.to_laurent_C_mul_T (n : ℕ) (r : R) :
+  ((polynomial.monomial n r).to_laurent : R[T;T⁻¹]) = C r * T n :=
+begin
+  show map_domain (nat.cast_add_monoid_hom ℤ) ((to_finsupp_iso R) (monomial n r)) =
+    (C r * T n : R[T;T⁻¹]),
+  convert (map_domain_single : _ = single (nat.cast_add_monoid_hom ℤ n) r),
+  { exact to_finsupp_monomial n r },
+  { simp only [nat.coe_cast_add_monoid_hom, int.nat_cast_eq_coe_nat, single_eq_C_mul_T] },
+end
+
+@[simp]
+lemma _root_.polynomial.to_laurent_C (r : R) :
+  (polynomial.C r).to_laurent = C r :=
+begin
+  convert polynomial.to_laurent_C_mul_T 0 r,
+  simp only [int.coe_nat_zero, T_zero, mul_one],
+end
+
+@[simp]
+lemma _root_.polynomial.to_laurent_X :
+  (polynomial.X.to_laurent : R[T;T⁻¹]) = T 1 :=
+begin
+  have : (polynomial.X : R[X]) = monomial 1 1,
+  { simp [monomial_eq_C_mul_X] },
+  simp [this, polynomial.to_laurent_C_mul_T],
+end
+
+@[simp] lemma _root_.polynomial.to_laurent_one : (polynomial.to_laurent : R[X] → R[T;T⁻¹]) 1 = 1 :=
+map_one polynomial.to_laurent
+
+@[simp]
+lemma _root_.polynomial.to_laurent_C_mul_eq (r : R) (f : R[X]):
+  (polynomial.C r * f).to_laurent = C r * f.to_laurent :=
+by simp only [_root_.map_mul, polynomial.to_laurent_C]
+
+@[simp]
+lemma _root_.polynomial.to_laurent_X_pow (n : ℕ) :
+  (X ^ n : R[X]).to_laurent = T n :=
+by simp only [map_pow, polynomial.to_laurent_X, T_pow, mul_one]
+
+@[simp]
+lemma _root_.polynomial.to_laurent_C_mul_X_pow (n : ℕ) (r : R) :
+  (polynomial.C r * X ^ n).to_laurent = C r * T n :=
+by simp only [_root_.map_mul, polynomial.to_laurent_C, polynomial.to_laurent_X_pow]
+
+lemma is_unit_T (n : ℤ) : is_unit (T n : R[T;T⁻¹]) :=
```

Comment:
@eric-wieser 
> Can you make the stronger statement `invertible (T n)`? You can then prove this one with `@is_unit_of_invertible _ _ T_invertible` or similar

Should I make `invertible_T` an instance or simply a definition?

Reply:
We're usually pretty cautious about making `invertible` declarations `instance`s, but this one seems pretty safe. Maybe @Vierkantor has an opinion.

Reply:
You could also consider making it a `localized` instance, like we do for `fact` instances. On the other hand, a global instance probably won't hurt (as long as nothing is going to try and unify `T` with something ugly).

Reply:
I'm not particularly attached to the instance.  I only proved it, since Eric suggested it: I had `is_unit` and I'm not even sure that I got to the point where I used it!